pytorch_to_returnn/import_wrapper/wrap.py

```diff
@@ -38,6 +38,8 @@ def wrap(obj, *, name: str, ctx: WrapCtx):
       pass  # keep as-is
     elif type(obj) != type:  # explicitly do not check sub-types, e.g. pybind11_type
       pass
+    elif issubclass(obj, tuple):  # e.g. namedtuple
+      pass  # keep as-is
```

Comment:
Why?

Reply:
To allow unpacking, e.g. `data, batch_sizes, sorted_indices, unsorted_indices = packed_seq`.

Reply:
I don't understand:

- Why or in what situation do we even call this function for a `PackedSequence`?
- Why does the unpacking not work for the wrapped object? This sounds like a bug which would be easy to fix, and which should be fixed anyway, and then this change here is not needed.
- Wouldn't this cause problems for other named tuples which are not over tensors?
- Doesn't this cause problems when the tensors inside the namedtuple are not wrapped then?

Reply:
> Why or in what situation do we even call this function for a PackedSequence?

This is called in `_run_traced_orig_torch()` when a `PackedSequence` is input to some module, like `LSTM` in `test_lstm_with_packed_sequence_input`. For the `LSTM` module, we run into a problem if the class is wrapped, because in the `forward`, there is a check `if isinstance(orig_input, PackedSequence): input, batch_sizes, sorted_indices, unsorted_indices = input`. For the wrapped object, the unpacking is skipped which is why the subsequent code crashes.

> Why does the unpacking not work for the wrapped object? This sounds like a bug which would be easy to fix, and which should be fixed anyway, and then this change here is not needed.

Actually, the unpacking works without the marked change. It was probably some other issue before.

> Wouldn't this cause problems for other named tuples which are not over tensors?
> Doesn't this cause problems when the tensors inside the namedtuple are not wrapped then?

This here is also not only over tensors, but so far always includes `None`. I didn't encounter issues because of non-wrapped tensors. In what situation could that occur?

Reply:
> > Why or in what situation do we even call this function for a PackedSequence?
> 
> This is called in `_run_traced_orig_torch()` when a `PackedSequence` is input to some module, like `LSTM` in `test_lstm_with_packed_sequence_input`.

Ah, I see now.

> For the `LSTM` module, we run into a problem if the class is wrapped, because in the `forward`, there is a check `if isinstance(orig_input, PackedSequence): input, batch_sizes, sorted_indices, unsorted_indices = input`.

Ah, but this can be fixed. We just also need to implement `__instancecheck__` and maybe `__subclasscheck__` in the wrapped object (in `WrappedObject`).

Also, isn't `PackedSequence` here also itself wrapped (via `make_wrapped_class`)? If not, why not? If yes, then why doesn't it work already? In `make_wrapped_object`, it would specifically create a wrapped object of the right type, so in principle, this `isinstance` check should already work.

> > Why does the unpacking not work for the wrapped object? This sounds like a bug which would be easy to fix, and which should be fixed anyway, and then this change here is not needed.
> 
> Actually, the unpacking works without the marked change. It was probably some other issue before.
> 
> > Wouldn't this cause problems for other named tuples which are not over tensors?
> > Doesn't this cause problems when the tensors inside the namedtuple are not wrapped then?
> 
> This here is also not only over tensors, but so far always includes `None`. I didn't encounter issues because of non-wrapped tensors. In what situation could that occur?

When you don't wrap `PackedSequence` (or other tuples or objects over tensors), then it would also not wrap the tensor inside. At least not here. Maybe it would wrap it by some follow-up code when the tensor is directly used. But not sure. Or maybe it doesn't cause problems here because in our Torch implementation, it would not exist like this.

However, this is now very specific to `PackedSequence`. In general, when a tuple or namedtuple contains tensors, those should be wrapped.

What we could do is to not wrap tuple, namedtuple etc but unpack them, wrap all their content, and then repack them. However, I'm not sure if this is better, or really needed. In principle, wrapping a tuple directly should also work.

Reply:
> Also, isn't PackedSequence here also itself wrapped (via make_wrapped_class)? If not, why not? If yes, then why doesn't it work already? In make_wrapped_object, it would specifically create a wrapped object of the right type, so in principle, this isinstance check should already work.

Yes, as I wrote in the edit above: In the padertorch example it works and I'm not sure why there is this difference. In the padertorch example, this function gets called for `PackedSequence` already a couple of times when padertorch is imported, maybe this makes the difference.

Reply:
> but then the unpacking fails (`TypeError: cannot unpack non-iterable PackedSequence object`).

So `isinstance(orig_input, PackedSequence)` returned True, and then the unpacking fails with this error? It's strange that it works in some cases but not in others.

But in any case, the unpacking of a `WrappedObject` is probably just not implemented right. I assume the dynamic `__getattr__` logic does not work, i.e. will not be called for other magic methods like `__len__`, `__iter__`, etc.

So maybe we should add such code to `WrappedObject.__init__`:
```
for special in ["__len__", "__iter__"]:
  if hasattr(orig_obj, special):
    setattr(self, getattr(self, special))  # make explicitly available
```

Reply:
It seems like this doesn't change anything. But the problem seems to be related. At least, I cannot call `len()` on the `PackedSequence`

```
>>> len(input)
*** TypeError: object of type 'PackedSequence' has no len()
>>> input.__len__()
4
```

Reply:
Ah yes, it doesn't work like that. It needs to be already assigned on the class, not on the object.

I think you need to add similar logic to `make_wrapped_class`. Sth like:
```
for special in ["__len__", "__iter__"]:
  if hasattr(cls, special):
    func = getattr(cls, special)
    wrapped_func = wrap(func, name="%s.%s" % (name, special), ctx=ctx)
    setattr(WrappedClass, wrapped_func)  # make explicitly available
```

Reply:
Yes, this seems to work with some small modifications :+1:

Reply:
What remains are these points

> > For the `LSTM` module, we run into a problem if the class is wrapped, because in the `forward`, there is a check `if isinstance(orig_input, PackedSequence): input, batch_sizes, sorted_indices, unsorted_indices = input`.
>
> Ah, but this can be fixed. We just also need to implement `__instancecheck__` and maybe `__subclasscheck__` in the wrapped object (in `WrappedObject`).
>
> Also, isn't `PackedSequence` here also itself wrapped (via `make_wrapped_class`)? If not, why not? If yes, then why doesn't it work already? In `make_wrapped_object`, it would specifically create a wrapped object of the right type, so in principle, this `isinstance` check should already work.
>
> > > Why does the unpacking not work for the wrapped object? This sounds like a bug which would be easy to fix, and which should be fixed anyway, and then this change here is not needed.
> > 
> > Actually, the unpacking works without the marked change. It was probably some other issue before.
> > 
> > > Wouldn't this cause problems for other named tuples which are not over tensors?
> > > Doesn't this cause problems when the tensors inside the namedtuple are not wrapped then?
> > 
> > This here is also not only over tensors, but so far always includes `None`. I didn't encounter issues because of non-wrapped tensors. In what situation could that occur?
>
> When you don't wrap `PackedSequence` (or other tuples or objects over tensors), then it would also not wrap the tensor inside. At least not here. Maybe it would wrap it by some follow-up code when the tensor is directly used. But not sure. Or maybe it doesn't cause problems here because in our Torch implementation, it would not exist like this.
>
> However, this is now very specific to `PackedSequence`. In general, when a tuple or namedtuple contains tensors, those should be wrapped.
>
> What we could do is to not wrap tuple, namedtuple etc but unpack them, wrap all their content, and then repack them. However, I'm not sure if this is better, or really needed. In principle, wrapping a tuple directly should also work.

Reply:
The difference is that in the padertorch example, we have something like

```
h = torch.nn.utils.rnn.pack_padded_sequence(inputs, [n_time] * n_batch)
...
h = PackedSequence(h.data, h.batch_sizes)
output, _ = torch.nn.LSTM(n_feat, n_feat)(h)
```

while in our test case, there is no direct initialization of the `PackedSequence`. This way, in our test `h` is of the original type, not the wrapped one.

Reply:
This here should also have a small comment. Or maybe just refer to the comment below.